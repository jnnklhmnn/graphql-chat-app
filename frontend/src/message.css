.message {
  padding: 20px 10px;
  text-align: right;
  display: flex;
  justify-content: space-between;
}

.avatar-wrapper {
  min-width: 50px;
}

.avatar {
  border-radius: 50%;
}

.edit-button {
  background-color: #a29a9a;
  cursor: pointer;
  box-shadow: none;
  border: none;
}

.edit-button:focus {
  outline: none
}

.edit-textarea {
  background-color: #a29a9a;
  resize: none;
  padding: 10px;
  text-align: left;
  line-height: 1.2;
  box-shadow: none;
  border: none;
  font-size: 16px;
  font-family: 'Raleway', sans-serif;
}

.edit-wrapper {
  display: flex;
  background-color: #a29a9a;
}

.message-body-wrapper {
  display: flex;
  align-items: baseline;
  background-color: #a29a9a;
  width: 100%;
  box-shadow: 0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset;
}

.message-copy {
  padding: 10px;
  text-align: left;
  line-height: 1.2;
  font-family: 'Raleway', sans-serif;
}

.message-body {
  margin-left: 10px;
}

.date {
    font-family: 'Raleway', sans-serif;
    padding-top: 8px;
    font-size: 12px;
    color: #635b5b;
}