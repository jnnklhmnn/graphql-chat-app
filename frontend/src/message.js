import React, { Component } from "react";
import PropTypes from "prop-types";
import { graphql } from "react-apollo";
import gql from "graphql-tag";
import DeleteButton from "./delete-button.js";
import { ALL_POSTS_QUERY } from "./App.js";
import "./message.css";

export class Message extends Component {
  constructor(props) {
    super(props);

    this.state = {
      message: this.props.message,
      editMode: false
    };
  }

  /**
   * Formats Date from GraphQL Timestamp (example:2018-05-06T20:34:04.000Z)
   * to DD/MM/YYYY HH:MM Format
   *
   * @param String date
   * @returns Markup including formated Date String
   *
   */

  getFormattedDate = date => {
    const year = date.substring(0, 4);
    const month = date.substring(5, 7);
    const day = date.substring(8, 10);
    const hours = date.substring(11, 13);
    const minutes = date.substring(14, 16);

    return (
      <div className="date">
        <p>
          {day}/{month}/{year} {hours}:{minutes}
        </p>
      </div>
    );
  };

  /**
   * Sends request to mutate the Schema to edit existing Message in Database,
   * triggers refetch of ALL_POSTS_QUERY.
   */

  editMessage = async () => {
    const { id } = this.props;
    const message = this.state.message;
    this.setState({ editMode: false });
    await this.props.editMutation({
      variables: {
        id,
        message
      },
      refetchQueries: [{ query: ALL_POSTS_QUERY }]
    });
  };

  /**
   * Creates the Markup for the avatar Placeholder
   *
   * @returns Markup
   */

  getAvatarPlaceholder = () => {
    return (
      <img
        className="avatar"
        alt="placeholder-avatar"
        src="http://via.placeholder.com/50"
        srcSet="http://via.placeholder.com/50 1x, http://via.placeholder.com/100 2x, http://via.placeholder.com/150 3x"
      />
    );
  };

  /**
   * Creates the Markup for message Section depending on the state in editMode
   * it returns the markup for the read or the write version of the interface
   *
   * @returns Markup
   */

  getMessageMode = () => {
    const Date = this.props.updatedAt
      ? this.getFormattedDate(this.props.updatedAt)
      : this.getFormattedDate(this.props.createdAt);
    if (this.state.editMode) {
      return (
        <div className="edit-wrapper">
          <textarea
            className="edit-textarea"
            type="text"
            value={this.state.message}
            onChange={e => this.setState({ message: e.target.value })}
          />
          <button className="edit-button" onClick={this.editMessage}>
            <svg
              width="15"
              height="15"
              viewBox="0 0 1792 1792"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path d="M1671 566q0 40-28 68l-724 724-136 136q-28 28-68 28t-68-28l-136-136-362-362q-28-28-28-68t28-68l136-136q28-28 68-28t68 28l294 295 656-657q28-28 68-28t68 28l136 136q28 28 28 68z" />
            </svg>
          </button>
        </div>
      );
    } else {
      return (
        <div className="message-body">
          <div className="message-body-wrapper">
            <p className="message-copy">{this.state.message}</p>
            <button
              className="edit-button"
              onClick={e => this.setState({ editMode: true })}
            >
              <svg
                width="15"
                height="15"
                viewBox="0 0 1792 1792"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path d="M491 1536l91-91-235-235-91 91v107h128v128h107zm523-928q0-22-22-22-10 0-17 7l-542 542q-7 7-7 17 0 22 22 22 10 0 17-7l542-542q7-7 7-17zm-54-192l416 416-832 832h-416v-416zm683 96q0 53-37 90l-166 166-416-416 166-165q36-38 90-38 53 0 91 38l235 234q37 39 37 91z" />
              </svg>
            </button>
            <DeleteButton id={this.props.id} />
          </div>
          {Date}
        </div>
      );
    }
  };

  render() {
    const MessageMode = this.getMessageMode();
    const avatar = this.getAvatarPlaceholder();

    return (
      <div className="message">
        <div className="avatar-wrapper">{avatar}</div>
        <div className="message-wrapper">{MessageMode}</div>
      </div>
    );
  }
}

const EDIT_MUTATION = gql`
  mutation($id: ID!, $message: String) {
    updatePost(message: $message, id: $id) {
      id
      updatedAt
    }
  }
`;

export default graphql(EDIT_MUTATION, { name: "editMutation" })(Message);

Message.propTypes = {
  message: PropTypes.string,
  createdAt: PropTypes.string,
  updatedAt: PropTypes.string
};
