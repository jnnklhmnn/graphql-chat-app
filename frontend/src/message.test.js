import { Message } from './message.js';
import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme from 'enzyme';

describe('Message', () => {
    test('renders with data correctly', () => {
      Enzyme.configure({ adapter: new Adapter() });
      const wrapper = shallow(<Message  message="Example Message" createdAt="2018-04-07T20:34:04.000Z" updatedAt="2018-05-06T20:34:04.000Z" />);
    expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });