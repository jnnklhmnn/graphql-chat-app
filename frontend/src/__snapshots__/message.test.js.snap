// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message renders with data correctly 1`] = `
<div
  className="message"
>
  <div
    className="avatar-wrapper"
  >
    <img
      alt="placeholder-avatar"
      className="avatar"
      src="http://via.placeholder.com/50"
      srcSet="http://via.placeholder.com/50 1x, http://via.placeholder.com/100 2x, http://via.placeholder.com/150 3x"
    />
  </div>
  <div
    className="message-wrapper"
  >
    <div
      className="message-body"
    >
      <div
        className="message-body-wrapper"
      >
        <p
          className="message-copy"
        >
          Example Message
        </p>
        <button
          className="edit-button"
          onClick={[Function]}
        >
          <svg
            height="15"
            viewBox="0 0 1792 1792"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M491 1536l91-91-235-235-91 91v107h128v128h107zm523-928q0-22-22-22-10 0-17 7l-542 542q-7 7-7 17 0 22 22 22 10 0 17-7l542-542q7-7 7-17zm-54-192l416 416-832 832h-416v-416zm683 96q0 53-37 90l-166 166-416-416 166-165q36-38 90-38 53 0 91 38l235 234q37 39 37 91z"
            />
          </svg>
        </button>
        <Apollo(DeleteButton) />
      </div>
      <div
        className="date"
      >
        <p>
          06
          /
          05
          /
          2018
           
          20
          :
          34
        </p>
      </div>
    </div>
  </div>
</div>
`;
